# ML-02 — Research Question and Provisional Lane

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [2]:
# My provisional lane is CTR / Engagement Opportunity Scoring. For the first version of the project, I will focus mainly on CTR opportunity scoring and use engagement metrics as supporting context where enough data is available.
#I selected this lane because it connects directly to a practical resource-allocation decision: when a team has limited time, which visible pages should it investigate first for possible CTR or engagement improvements?

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

'''
Research question

Among pages with sufficient search impressions, which pages have a lower CTR than expected after accounting for their average search position and search volume, and should therefore be prioritized for human review?

Decision being improved

The decision is which pages should enter a limited CTR and engagement review queue first.

Output

The final project should produce a ranked opportunity score for each eligible page. A higher score should indicate stronger evidence that the page is under-capturing clicks relative to comparable pages.

The score should consider factors such as:

impressions;
CTR;
average position;
expected CTR for the page's position tier;
the size of the CTR gap;
sessions and engagement context;
content age, freshness or content type where useful.
Action someone could take

A content or search specialist could review the highest-ranked pages and decide whether to:

improve the page title or meta description;
improve alignment with search intent;
improve snippet-friendly page structure;
strengthen the opening section or on-page content;
investigate weak engagement;
leave the page unchanged and continue monitoring it.

The score itself will not automatically prescribe an edit. It will identify pages that deserve investigation.

Cost of a wrong recommendation

A false positive could waste reviewer time or lead someone to change a page that was already performing appropriately. An unnecessary title or content change could even reduce performance.

A false negative could cause the team to overlook a page where a realistic opportunity to gain additional clicks exists.

For this reason, the system should prioritize explainable recommendations, sufficient data volume and human review rather than treating every low CTR value as a problem.

Why data or machine learning may help

CTR depends strongly on average ranking position and may also vary with visibility, intent, content type, freshness and other page characteristics. A data-driven method can create fairer comparisons than using one universal CTR threshold.

Machine learning may later help detect interactions between these signals, but the project must first establish a transparent position-adjusted baseline. A more complicated model will only be useful if it improves the quality of the ranked review list during proper validation.
'''

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [6]:
%pip install numpy pandas

Defaulting to user installation because normal site-packages is not writeable
   ---------------------------------------- 0.0/12.4 MB ? eta -:--:--
   ---------------------------------------- 0.0/12.4 MB ? eta -:--:--
   - -------------------------------------- 0.5/12.4 MB 2.4 MB/s eta 0:00:05
   ----------- ---------------------------- 3.7/12.4 MB 8.7 MB/s eta 0:00:02
   -------------------------- ------------- 8.1/12.4 MB 15.7 MB/s eta 0:00:01
   ---------------------------- ----------- 8.9/12.4 MB 11.8 MB/s eta 0:00:01
   ------------------------------ --------- 9.4/12.4 MB 11.1 MB/s eta 0:00:01
   ----------------------------------- ---- 11.0/12.4 MB 9.1 MB/s eta 0:00:01
   ------------------------------------- -- 11.5/12.4 MB 8.0 MB/s eta 0:00:01
   ---------------------------------------- 12.4/12.4 MB 7.9 MB/s eta 0:00:00
   ---------------------------------------- 0.0/9.8 MB ? eta -:--:--
   -- ------------------------------------- 0.5/9.8 MB 2.4 MB/s eta 0:00:04
   ----- ------


[notice] A new release of pip is available: 25.0.1 -> 26.1.2
[notice] To update, run: C:\Users\Ahmed\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\python.exe -m pip install --upgrade pip


In [7]:
from pathlib import Path

import numpy as np
import pandas as pd

# This handles common notebook working-directory locations.
possible_paths = [
    Path("data/raw/content_refresh_anonymized.csv"),
    Path("../../data/raw/content_refresh_anonymized.csv"),
    Path("../data/raw/content_refresh_anonymized.csv"),
]

data_path = next(
    (path for path in possible_paths if path.exists()),
    None,
)

if data_path is None:
    raise FileNotFoundError(
        "Could not find data/raw/content_refresh_anonymized.csv. "
        "Check where the notebook kernel is running."
    )

df = pd.read_csv(data_path)

print(f"Loaded: {data_path.resolve()}")
print(f"Rows: {len(df):,}")
print(f"Columns: {len(df.columns):,}")

df.head()

Loaded: C:\Users\Ahmed\flyranks_week1_notebook_01\data\raw\content_refresh_anonymized.csv
Rows: 30,000
Columns: 44


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [8]:
required_columns = [
    "content_id",
    "impressions_90d",
    "clicks_90d",
    "ctr",
    "avg_position",
]

missing_columns = [
    column for column in required_columns
    if column not in df.columns
]

if missing_columns:
    raise ValueError(
        f"Missing required columns: {missing_columns}"
    )

df[required_columns].describe().T

,count,mean,std,min,25%,50%,75%,max
impressions_90d,30000.0,5200.366300,16838.019547,1.0,81.0,731.00,3615.25,517715.0
clicks_90d,30000.0,16.097333,75.076958,0.0,0.0,1.00,7.00,4178.0
ctr,30000.0,0.510733,3.279162,0.0,0.0,0.07,0.29,100.0
avg_position,30000.0,16.342380,15.216790,0.0,6.2,10.80,22.30,245.0


In [9]:
analysis_df = df.copy()

numeric_columns = [
    "impressions_90d",
    "clicks_90d",
    "ctr",
    "avg_position",
]

for column in numeric_columns:
    analysis_df[column] = pd.to_numeric(
        analysis_df[column],
        errors="coerce",
    )

# Remove duplicate content pages for a page-level analysis.
analysis_df = analysis_df.drop_duplicates(
    subset="content_id"
).copy()

# Convert CTR to a rate between 0 and 1 when it appears to be stored
# as a percentage between 0 and 100.
ctr_99th_percentile = analysis_df["ctr"].quantile(0.99)

if ctr_99th_percentile > 1:
    analysis_df["ctr_rate"] = analysis_df["ctr"] / 100
    ctr_format = "percentage"
else:
    analysis_df["ctr_rate"] = analysis_df["ctr"]
    ctr_format = "decimal rate"

print(f"Detected CTR format: {ctr_format}")
print(f"Unique content pages: {len(analysis_df):,}")
print(
    f"Median CTR: "
    f"{analysis_df['ctr_rate'].median():.2%}"
)

Detected CTR format: percentage
Unique content pages: 30,000
Median CTR: 0.07%


In [10]:
eligible = analysis_df.loc[
    (analysis_df["impressions_90d"] >= 500)
    & (analysis_df["avg_position"] > 0)
    & (analysis_df["avg_position"] <= 20)
    & analysis_df["ctr_rate"].notna()
].copy()

print(f"All unique pages: {len(analysis_df):,}")
print(f"Eligible visible pages: {len(eligible):,}")
print(
    "Eligible percentage: "
    f"{len(eligible) / len(analysis_df):.2%}"
)

All unique pages: 30,000
Eligible visible pages: 12,023
Eligible percentage: 40.08%


In [11]:
analysis_df = df.copy()

numeric_columns = [
    "impressions_90d",
    "clicks_90d",
    "ctr",
    "avg_position",
]

for column in numeric_columns:
    analysis_df[column] = pd.to_numeric(
        analysis_df[column],
        errors="coerce",
    )

# Remove duplicate content pages for a page-level analysis.
analysis_df = analysis_df.drop_duplicates(
    subset="content_id"
).copy()

# Convert CTR to a rate between 0 and 1 when it appears to be stored
# as a percentage between 0 and 100.
ctr_99th_percentile = analysis_df["ctr"].quantile(0.99)

if ctr_99th_percentile > 1:
    analysis_df["ctr_rate"] = analysis_df["ctr"] / 100
    ctr_format = "percentage"
else:
    analysis_df["ctr_rate"] = analysis_df["ctr"]
    ctr_format = "decimal rate"

print(f"Detected CTR format: {ctr_format}")
print(f"Unique content pages: {len(analysis_df):,}")
print(
    f"Median CTR: "
    f"{analysis_df['ctr_rate'].median():.2%}"
)

Detected CTR format: percentage
Unique content pages: 30,000
Median CTR: 0.07%


In [13]:
eligible["expected_ctr_for_tier"] = (
    eligible.groupby(
        "position_tier",
        observed=True,
    )["ctr_rate"]
    .transform("median")
)

eligible["ctr_gap"] = (
    eligible["expected_ctr_for_tier"]
    - eligible["ctr_rate"]
)

eligible["below_tier_median"] = (
    eligible["ctr_gap"] > 0
)

underperforming = eligible.loc[
    eligible["below_tier_median"]
].copy()

underperforming["directional_click_gap"] = (
    underperforming["ctr_gap"]
    * underperforming["impressions_90d"]
)

print(
    "Pages below their position-tier median CTR: "
    f"{len(underperforming):,}"
)

print(
    "Percentage of eligible pages below benchmark: "
    f"{len(underperforming) / len(eligible):.2%}"
)

print(
    "Directional click-gap estimate: "
    f"{underperforming['directional_click_gap'].sum():,.0f}"
)

Pages below their position-tier median CTR: 5,888
Percentage of eligible pages below benchmark: 48.97%
Directional click-gap estimate: 53,633


In [14]:
candidate_columns = [
    "content_id",
    "impressions_90d",
    "clicks_90d",
    "avg_position",
    "position_tier",
    "ctr_rate",
    "expected_ctr_for_tier",
    "ctr_gap",
    "directional_click_gap",
]

top_candidates = (
    underperforming[candidate_columns]
    .sort_values(
        "directional_click_gap",
        ascending=False,
    )
    .head(10)
    .copy()
)

percentage_columns = [
    "ctr_rate",
    "expected_ctr_for_tier",
    "ctr_gap",
]

for column in percentage_columns:
    top_candidates[column] = (
        top_candidates[column] * 100
    ).round(2)

top_candidates = top_candidates.rename(
    columns={
        "ctr_rate": "ctr_pct",
        "expected_ctr_for_tier": "expected_ctr_pct",
        "ctr_gap": "ctr_gap_percentage_points",
    }
)

top_candidates

,content_id,impressions_90d,clicks_90d,avg_position,position_tier,ctr_pct,expected_ctr_pct,ctr_gap_percentage_points,directional_click_gap
3394,content_36ff89c8214e,295097,154,7.3,page_1,0.05,0.24,0.19,560.6843
6653,content_5fe46e04994d,517715,741,4.2,page_1,0.14,0.24,0.10,517.7150
7445,content_c8e9d6ab9013,208678,0,9.7,page_1,0.00,0.24,0.24,500.8272
6903,content_c84a0ab98e90,223271,70,7.8,page_1,0.03,0.24,0.21,468.8691
7678,content_8451fc6f034d,272144,75,2.3,top_3,0.03,0.20,0.17,462.6448
27178,content_453722754fea,140079,16,7.6,page_1,0.01,0.24,0.23,322.1817
22028,content_73c54f78c06a,213963,211,4.7,page_1,0.10,0.24,0.14,299.5482
3070,content_91652435f57a,159590,100,7.8,page_1,0.06,0.24,0.18,287.2620
9193,content_c1fe78bc4e37,134055,43,7.5,page_1,0.03,0.24,0.21,281.5155
15914,content_0919dd345d80,119217,26,7.0,page_1,0.02,0.24,0.22,262.2774


In [15]:
summary_numbers = {
    "unique_pages": len(analysis_df),
    "eligible_pages": len(eligible),
    "eligible_share": len(eligible) / len(analysis_df),
    "below_benchmark_pages": len(underperforming),
    "below_benchmark_share": (
        len(underperforming) / len(eligible)
    ),
    "directional_click_gap": (
        underperforming["directional_click_gap"].sum()
    ),
}

summary_numbers

{'unique_pages': 30000,
 'eligible_pages': 12023,
 'eligible_share': 0.40076666666666666,
 'below_benchmark_pages': 5888,
 'below_benchmark_share': 0.4897280212925227,
 'directional_click_gap': np.float64(53633.277799999996)}

## 3. Quick look at the data

The starter dataset contains **30000 unique anonymized content pages**.

After applying an initial reliability rule of at least 500 impressions and an average position between 1 and 20, **12023 pages remained**, representing approximately **12023%** of the available pages.

Within this eligible group, **5888 pages**, or approximately **48.97%%**, had a CTR below the median CTR for their own position tier.

The initial position-adjusted calculation produced a directional click gap of approximately **53633.277799999996 clicks** across these pages. This figure is useful for ranking potential opportunities, but it should not be interpreted as the number of clicks that would definitely be recovered through changes.

These numbers suggest that the dataset contains enough visible pages and enough variation in position-adjusted CTR to support further investigation over the next seven weeks. They also show why the project should compare pages within similar ranking positions rather than marking every page below one universal CTR threshold as underperforming.


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

### What this project may be able to claim

- A page's observed CTR is lower than the typical CTR of pages in a comparable position tier.
- A page has enough impressions for its CTR gap to deserve investigation.
- A ranking method can help prioritize a limited human review queue.
- Some observable page characteristics may be associated with CTR or engagement differences.
- A tested method may rank review candidates better than a simple transparent baseline.

### What this project cannot currently claim

- That a title or meta-description change will cause CTR to increase.
- That low CTR is always caused by poor metadata.
- That every estimated click gap is recoverable.
- That the identified features are Google ranking factors.
- That a high opportunity score means a page should automatically be changed.
- That observational data proves causation.

The recommendations will therefore be described as review opportunities rather than guaranteed improvements. Human inspection will remain necessary before any action is taken.

## Self-check

Before you submit, confirm each line honestly:

- [✔] Every section above is filled — markdown thinking AND the code that backs it
- [✔] The notebook runs top to bottom with no errors (Runtime → Run all)
- [✔] No client names, URLs, or private queries anywhere
- [✔] My claims use careful words: observed, measured, directional, decision-support
- [✔] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.